R/makeTreeSummarizedExperimentFromqiime2.R

```diff
@@ -0,0 +1,245 @@
+    } else {
+        stop(
+            "Only files in format of 'BIOMV210DirFmt', ",
+            "'TSVTaxonomyDirectoryFormat', NewickDirectoryFormat' and ",
+            "'DNASequencesDirectoryFormat' are supported."
+        )
+    }
+
+    res
+}
+
+
+#' Read qiime2 feature table
+#'
+#' Import qiime2 feature table into otu_table object
+#'
+#' @param file character, file name of the biom file.
+#' @return [`phyloseq::otu_table-class`] object.
+#' @noRd
+#' @keywords internal
+read_q2biom <- function(file) {
+    biomobj <- suppressWarnings(biomformat::read_biom(file))
+    feature_tab <- as(biomformat::biom_data(biomobj),"matrix")
+
+    feature_tab
+}
+
+#' Read qiime2 taxa file
+#' @keywords internal
+#' @noRd
+read_q2taxa <- function(file) {
+    taxa <- utils::read.table(file, sep = '\t', header = TRUE)
+    if ("Confidence" %in% names(taxa)) {
+        taxa$Confidence <- NULL
+    }
+
+    as.matrix(taxa)
+}
+
+#' Read qiime2 sample meta data file
+#' @keywords internal
+#' @noRd
+read_q2sample_meta <- function(file) {
+    sam <- utils::read.table(file = file, header = TRUE, sep = "\t", comment.char = "")
+    rownames(sam) <- as.character(sam[, 1])
+
+    # remove row: #q2:types
+    idx <- which(sam == "#q2:types", arr.ind = TRUE)
+    sam <- sam[-idx[, "row"], ]
+
+    S4Vectors::DataFrame(sam)
+}
+
+
+#' Subset taxa according to the taxa in feature table
+#' @keywords internal
+#' @noRd
+subset_taxa_in_feature <- function(taxa_tab, feature_tab) {
+    idx <- match(row.names(feature_tab), taxa_tab[, "Feature.ID"])
+    taxa_tab <- taxa_tab[idx, , drop = FALSE]
+    row.names(taxa_tab) <- taxa_tab[, "Feature.ID"]
+
+    taxa_tab
+}
+
+#' Parse qiime2 taxa in different taxonomic levels
+#' @param taxa `tax_table` object.
+#' @param sep character string containing a regular expression, separator
+#'  between different taxonomic levels, defaults to on compatible with both
+#'  GreenGenes and SILVA `; |;"`.
+#' @param trim_rank_prefix logical whether remove leading characters from
+#'   taxonomic levels, e.g. k__ or D_0__, default `TRUE`.
+#' @return [`phyloseq::taxonomyTable-class`] object.
+#' @keywords internal
+#' @importFrom rlang .data
+#' @noRd
+parse_q2taxonomy <- function(taxa, sep = "; |;", trim_rank_prefix = TRUE) {
+    taxa <- data.frame(taxa)
+    if(trim_rank_prefix){
+        # remove leading characters from GG
+        taxa$Taxon <- gsub("[kpcofgs]__", "", taxa$Taxon)
+        #remove leading characters from SILVA
+        taxa$Taxon <- gsub("D_\\d__", "", taxa$Taxon)
+    }
+
+    taxa <- suppressWarnings(
+        tidyr::separate(
+            taxa, .data$Taxon,
+            c("Kingdom","Phylum","Class","Order","Family","Genus","Species"),
+            sep = sep
+        )
+    )
+    taxa <- sapply(taxa, function(x) ifelse(x == "", NA_character_, x))
+    rownames(taxa) <- taxa[, "Feature.ID"]
+    taxa <- taxa[, setdiff(colnames(taxa), "Feature.ID")]
+
+    taxa
+}
```

Comment:
This function works only, if all taxonomic levels are present.

Currently there is no check whether all are present. Refactor accordingly or refactor to work with any combination of taxonomic levels available.